src/transformers/models/opt/modeling_opt.py

```diff
@@ -0,0 +1,1821 @@
+        dropout: float = 0.1,
+        is_decoder: bool = True,
+        bias: bool = True,
+        add_zero_attn = False
+    ):
+        super().__init__()
+        self.embed_dim = embed_dim
+        self.num_heads = num_heads
+        self.dropout = dropout
+        self.head_dim = embed_dim // num_heads
+        self.add_zero_attn = add_zero_attn
+
+        if (self.head_dim * num_heads) != self.embed_dim:
+            raise ValueError(
+                f"embed_dim must be divisible by num_heads (got `embed_dim`: {self.embed_dim}"
+                f" and `num_heads`: {num_heads})."
+            )
+        self.scaling = self.head_dim**-0.5
+        self.is_decoder = is_decoder
+
+        self.k_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.v_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.q_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+        self.out_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
+
+    def _shape(self, tensor: torch.Tensor, seq_len: int, bsz: int):
+        return tensor.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2).contiguous()
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        key_value_states: Optional[torch.Tensor] = None,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        layer_head_mask: Optional[torch.Tensor] = None,
+        output_attentions: bool = False,
+        self_attn_padding_mask: Optional[Tensor] = None,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        """Input shape: Batch x Time x Channel"""
+
+        # if key_value_states are provided this layer is used as a cross-attention layer
+        # for the decoder
+        is_cross_attention = key_value_states is not None
+        residual = hidden_states
+
+        # create key padding mask
+        assert torch.equal(attention_mask[0, 0, :, :], attention_mask[-1, 0, :, :])
+        attention_mask = attention_mask[0, 0, :, :]
+
+        return F.multi_head_attention_forward(
+                hidden_states,
+                hidden_states,
+                hidden_states,
+                self.embed_dim,
+                self.num_heads,
+                torch.empty([0]),
+                torch.cat((self.q_proj.bias, self.k_proj.bias, self.v_proj.bias)),
+                None,
+                None,
+                self.add_zero_attn,
+                self.dropout,
+                self.out_proj.weight,
+                self.out_proj.bias,
+                self.training,
+                self_attn_padding_mask,
+                False,
+                attention_mask,
+                use_separate_proj_weight=True,
+                q_proj_weight=self.q_proj.weight,
+                k_proj_weight=self.k_proj.weight,
+                v_proj_weight=self.v_proj.weight,
+            )
+
+        # bsz, tgt_len, _ = hidden_states.size()
```

Comment:
@younesbelkada does it not work with the commented out code or is it just "cleaner" to use PyTorch's `multi_head_attention_forward`?

Reply:
It is what they used in their implementation and worked to obtain the same logits (also I found it a bit cleaner)

Reply:
resolved in 0484ca10e74a245cd878166e3384eb51a0d97042